signac/contrib/collection.py

```diff
@@ -494,7 +494,9 @@ def _update_indexes(self):
         if self._dirty:
             for _id in self._dirty:
                 self._remove_from_indexes(_id)
-            docs = [self[_id] for _id in self._dirty]
+            # For performance, we read directly from the internal _docs dict to
+            # avoid copying the data like in __getitem__.
+            docs = [self._docs[_id] for _id in self._dirty]
```

Comment:
I haven't looked at Collection code in a very long time so I can't be 100% sure without doing a more thorough re-review of the codebase, but changing this to a non-copying operation will cascade that reference through to `self._indexes`. Are we sure this change is safe? It seems plausible, just covering our bases.

Reply:
This is a good question. I know it's safe for signac's internal use of `Collection` (because we create the `Collection`, use it, and then throw it away). However, I will investigate this a little further to make sure. If it turns out to be unsafe in the general case, I think that this is an indication that we should consider a minimized form of `Collection` for signac's internal use with searching. There's a lot of extraneous logic and copying that could be eliminated.

Reply:
@vyasr If it's to your preference, I might just revert the changes to `Collection` and then make a follow-up PR that implements a minimal `Collection` class that we can use internally. It seems like that would be a better plan in the long term, anyway.

Reply:
I reverted the changes to `Collection` in a4b08dc42dea35cbfb3f636c93a4194ad39ecb90. I will create a separate issue to document the potential optimizations in that class (or a minimized version of the class).

Reply:
Issue #664 addresses this.

Reply:
Sounds good, I think this is the safer choice for now. Let's restrict this PR to just those changes you have now that remove largely redundant functions from project.py. `Collection` is complicated and something none of us have touched in a while, so I'd prefer to make a focused effort on improving that if we think it's worthwhile.